lib/ansible/modules/network/aci/msc_user.py

```diff
@@ -0,0 +1,231 @@
+
+    user_id = module.params['user_id']
+    user_name = module.params['user']
+    user_password = module.params['user_password']
+    first_name = module.params['first_name']
+    last_name = module.params['last_name']
+    email = module.params['email']
+    phone = module.params['phone']
+    account_status = module.params['account_status']
+    # FIXME: Look up domain
+    domain = module.params['domain']
+    # FIXME: Look up roles
+    roles = module.params['roles']
+    roles_dict = list()
+    if roles:
+        for role in roles:
+            roles_dict.append(dict(roleId=role))
+    state = module.params['state']
+
+    msc = MSCModule(module)
+
+    path = 'users'
+
+    # Query for existing object(s)
+    if user_id is None and user_name is None:
+        msc.existing = msc.query_objs(path)
+    elif user_id is None:
+        msc.existing = msc.get_obj(path, username=user_name)
+        if msc.existing:
+            user_id = msc.existing['id']
+    elif user_name is None:
+        msc.existing = msc.get_obj(path, id=user_id)
+    else:
+        msc.existing = msc.get_obj(path, id=user_id)
+        existing_by_name = msc.get_obj(path, username=user_name)
+        if existing_by_name and user_id != existing_by_name['id']:
+            msc.fail_json(msg="Provided user '{1}' with id '{2}' does not match existing id '{3}'.".format(user_name, user_id, existing_by_name['id']))
+
+    # If we found an existing object, continue with it
+    if user_id:
+        path = 'users/{id}'.format(id=user_id)
+
+    if state == 'query':
+        pass
+
+    elif state == 'absent':
+        msc.previous = msc.existing
+        if msc.existing:
+            if module.check_mode:
+                msc.existing = {}
+            else:
+                msc.existing = msc.request(path, method='DELETE')
+
+    elif state == 'present':
+        msc.previous = msc.existing
+
+        msc.sanitize(dict(
+            id=user_id,
+            username=user_name,
+            password=user_password,
+            firstName=first_name,
+            lastName=last_name,
+            emailAddress=email,
+            phoneNumber=phone,
+            # accountStatus={},
+            accountStatus=account_status,
+            needsPasswordUpdate=False,
+            domainId=domain,
+            roles=roles_dict,
+            # active=True,
+            # remote=True,
+        ), collate=True)
+
+        if msc.existing:
+            if not issubset(msc.sent, msc.existing):
+                # NOTE: Since MSC always returns '******' as password, we need to assume a change
+                if 'password' in msc.sent:
+                    msc.result['changed'] = True
+
+                if module.check_mode:
+                    msc.existing = msc.proposed
+                else:
+                    msc.existing = msc.request(path, method='PUT', data=msc.sent)
+        else:
+            if module.check_mode:
+                msc.existing = msc.proposed
+            else:
```

Comment:
Should this have an update to msc.changed?

Reply:
The functions `exit_json()` and `fail_json()` set changed when existing and previous are different.
So we don't have to take care of this ourselves.